Deduplicate CI and setup scripts.

arthurp opened the pull request:
This moves changes from enterprise `setup_ubuntu.sh` to the open repo.

arthurp committed:
Extract repeated scripts in CPP CI builds. Move setup scripts.

The scripts are not used just for CI, so move them into scripts/.

arthurp committed:
Extract repeated scripts from Python CI builds.

arthurp committed:
Setup ccache directory explicitly to remove macOS differences.

arthurp committed:
Claim TODOs

ddn0 commented on .github/workflows/check_docs_ci.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
```
When these snippets are run by github actions, they are run with `set -e`; we should probably at least do the same here.

If we want to be more paranoid, `set -euo pipefail` is a good default for most scripts. If you add `-u` you may have to do something like `${1-}` to ensure that `$1` is always defined. Or simply allow bash to handle that as an error since that's what happens below. I don't mind being terse here since the only CI workflows call these scripts.

ddn0 commented on .github/workflows/configure_cpp_ci.sh:
```diff
@@ -0,0 +1,84 @@
+#! /bin/bash
```
ditto about `set -e` and co, and `#!/bin/bash` for consistency

ddn0 commented on .github/workflows/cpp.yaml:
```diff
@@ -47,19 +47,11 @@ jobs:
-        conan install -if $HOME/build --build=missing config || \
-          (rm -rf ~/.conan && .github/workflows/setup_conan.sh && \
-             conan install -if $HOME/build --build=missing config)
+      run: .github/workflows/setup_conan_ci_install.sh $HOME/build
```
if we are going to add `ci` to names everywhere, can this be `setup_conan_install_ci.sh` so it matches the other scripts?

Or simply drop `ci` altogether since it can be understood from the placement of these files in `.github/workflows`.

If you are worried about the name `.github/workflows/check_docs.sh` shadowing `scripts/check_docs.sh` (and somewhat similarly for `setup_conan_ci_install.sh` and `setup_conan.sh`, you can put all the CI related stuff into one wrapper script, e.g., `.github/workflows/ci_wrap.sh check_docs ...; ` (etc.)

arthurp replied:
Since these directories are never added to `PATH` I'm not really worried about duplcate names. I'll go with your initial thought.

ddn0 commented on .github/workflows/download_miniconda.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
```
set -e

ddn0 commented on .github/workflows/check_docs_ci.sh:
```diff
@@ -0,0 +1,15 @@
+  exit 1
+fi
+
+CHECK_DOCS="$(cd "$(dirname "$0")"/../../scripts || exit 1; pwd)"/check_docs.sh
```
We don't add `|| exit 1` elsewhere and the script will fail anyway if `CHECK_DOCS` doesn't point to `check_docs.sh`. Perhaps drop?

I also have a slight preference for just having `REPO_ROOT="$(cd "$(dirname "$0")/../.."); pwd)"` at the top of all these scripts so everything can be referred to consistently.

(One can get pretty sophisticated/cryptic with getting the path [1, 2]. Anything simple and clear is fine with me. Anything that needs to be robust probably shouldn't be done in bash anyway.)


[1] https://stackoverflow.com/questions/59895/how-to-get-the-source-directory-of-a-bash-script-from-within-the-script-itself
[2] https://stackoverflow.com/questions/4774054/reliable-way-for-a-bash-script-to-get-the-full-path-to-itself

arthurp replied:
I'll take your simple consistent approach. I really wish the BSD runtime had `realpath`. Lacking it really makes scripting correctly harder.

ddn0 commented on .github/workflows/setup_conan_ci_install.sh:
```diff
@@ -0,0 +1,20 @@
+#! /bin/bash
```
#!/bin/bash

ddn0 commented on .github/workflows/setup_miniconda.sh:
```diff
@@ -0,0 +1,7 @@
+# This is a bash INCLUDE file. Source it don't run it.
```
maybe `activate_miniconda.sh` is a better name because this doesn't install or configure anything unlike the other `setup_` scripts.

arthurp commented:
@ddn0 I think I have addressed all your concerns. Thanks for the review. Lots of good advise.

BTW, I considered adding `set -x` to the bash scripts used in CI so that what is executed is obvious in the action log. Do you think that would be a good idea?

ddn0 commented:
> BTW, I considered adding set -x to the bash scripts used in CI so that what is executed is obvious in the action log. Do you think that would be a good idea?

Yeah `set -x` is a good idea.
